File: general_ledger_report/general_ledger_report/custom_general_ledger/custom_general_ledger.py
from __future__ import unicode_literals
import frappe
from frappe import _

def execute(filters=None):
    """
    Main execution function for the Custom General Ledger report.
    
    :param filters: Dict containing filter parameters
    :return: Tuple of columns and data
    """
    columns = get_columns()
    data = get_data(filters)
    return columns, data

def get_columns():
    """
    Define the columns for the report.
    
    :return: List of column definitions
    """
    return [
        {
            "fieldname": "posting_date",
            "label": _("Date"),
            "fieldtype": "Date",
            "width": 90
        },
        {
            "fieldname": "voucher_type",
            "label": _("Voucher Type"),
            "fieldtype": "Data",
            "width": 120
        },
        {
            "fieldname": "debit",
            "label": _("Debit"),
            "fieldtype": "Currency",
            "width": 130
        },
        {
            "fieldname": "credit",
            "label": _("Credit"),
            "fieldtype": "Currency",
            "width": 130
        }
    ]

def get_data(filters):
    """
    Retrieve and process data for the report based on filters.
    
    :param filters: Dict containing filter parameters
    :return: List of data rows
    """
    # Construct the SQL query
    query = """
        SELECT 
            posting_date,
            voucher_type,
            debit,
            credit
        FROM 
            `tabGL Entry`
        WHERE 
            company = %(company)s
            AND posting_date BETWEEN %(from_date)s AND %(to_date)s
        ORDER BY 
            posting_date
    """
    
    # Execute the query with filters
    result = frappe.db.sql(query, filters, as_dict=1)
    
    return result

File: general_ledger_report/general_ledger_report/custom_general_ledger/custom_general_ledger.js
frappe.query_reports["Custom General Ledger"] = {
    "filters": [
        {
            "fieldname": "company",
            "label": __("Company"),
            "fieldtype": "Link",
            "options": "Company",
            "default": frappe.defaults.get_user_default("Company"),
            "reqd": 1
        },
        {
            "fieldname": "from_date",
            "label": __("From Date"),
            "fieldtype": "Date",
            "default": frappe.datetime.add_months(frappe.datetime.get_today(), -1),
            "reqd": 1
        },
        {
            "fieldname": "to_date",
            "label": __("To Date"),
            "fieldtype": "Date",
            "default": frappe.datetime.get_today(),
            "reqd": 1
        }
    ],
    
    // Add a formatter for currency fields
    "formatter": function(value, row, column, data, default_formatter) {
        if (column.fieldtype == "Currency") {
            value = format_currency(value, "INR");
        }
        return value;
    },
    
    // Add a function to show total at the end of the report
    "onload": function(report) {
        report.page.add_inner_button(__("Show Total"), function() {
            var total_debit = 0;
            var total_credit = 0;
            
            // Calculate totals
            report.data.forEach(function(row) {
                total_debit += row.debit;
                total_credit += row.credit;
            });
            
            // Show totals in a dialog
            frappe.msgprint(
                __("Total Debit: {0}<br>Total Credit: {1}", [
                    format_currency(total_debit, "INR"),
                    format_currency(total_credit, "INR")
                ])
            );
        });
    }
};